File: packages/client/src/prediction/predictor.ts
import {
  SERVER_TICK_MS,
  applyCommandToState,
  stepLocalPlayer,
  type Command,
  type MoveConfig,
  type MoveState,
} from '@ac/shared';

export const PREDICTION_SUBSTEP_MS = SERVER_TICK_MS;
export const MAX_SUBSTEPS_PER_FRAME = 5;

export interface PredictedPosition {
  x: number;
  y: number;
  z: number;
}

export interface Predictor {
  readonly state: MoveState;
  readonly pendingMs: number;
  setConfig(config: MoveConfig): void;
  setAuthoritative(x: number, y: number, z: number, yaw: number, pitch: number): void;
  stepWith(command: Command | undefined): void;
  advance(dtMs: number, command: Command | undefined): number;
  renderPosition(out: PredictedPosition): PredictedPosition;
  reset(): void;
}

export function createPredictor(config: MoveConfig): Predictor {
  const state: MoveState = {
    pos: { x: 0, y: 0, z: 0 },
    vel: { x: 0, y: 0, z: 0 },
    yaw: 0,
    pitch: 0,
  };
  let activeConfig = config;
  let accumulatorMs = 0;

  return {
    state,
    get pendingMs(): number {
      return accumulatorMs;
    },
    setConfig(next: MoveConfig): void {
      activeConfig = next;
    },
    setAuthoritative(x: number, y: number, z: number, yaw: number, pitch: number): void {
      state.pos.x = x;
      state.pos.y = y;
      state.pos.z = z;
      state.vel.x = 0;
      state.vel.y = 0;
      state.vel.z = 0;
      state.yaw = yaw;
      state.pitch = pitch;
      // 刻意不清零 accumulatorMs：它是"渲染节拍余量"，不是模拟状态。
      // 清零会让 20Hz 和解后的两帧渲染位置原地不动（相机每 3 帧才挪一次 = 画面一顿一顿）。
    },
    stepWith(command: Command | undefined): void {
      stepLocalPlayer(state, command, activeConfig, PREDICTION_SUBSTEP_MS);
    },
    advance(dtMs: number, command: Command | undefined): number {
      if (!Number.isFinite(dtMs) || dtMs <= 0) return 0;
      accumulatorMs += Math.min(dtMs, PREDICTION_SUBSTEP_MS * MAX_SUBSTEPS_PER_FRAME);
      let steps = 0;
      while (accumulatorMs >= PREDICTION_SUBSTEP_MS && steps < MAX_SUBSTEPS_PER_FRAME) {
        stepLocalPlayer(state, command, activeConfig, PREDICTION_SUBSTEP_MS);
        accumulatorMs -= PREDICTION_SUBSTEP_MS;
        steps += 1;
      }
      if (accumulatorMs > PREDICTION_SUBSTEP_MS) accumulatorMs = PREDICTION_SUBSTEP_MS;
      // 本帧没攒够子步时也要把渲染速度刷成当前意图：renderPosition 用 state.vel 做子步内线性外推，
      // 速度为 0 的帧渲染位置完全不动（顿挫的另一半根因）。
      if (steps === 0) applyCommandToState(state, command, activeConfig.player);
      return steps;
    },
    renderPosition(out: PredictedPosition): PredictedPosition {
      const seconds = accumulatorMs / 1000;
      out.x = state.pos.x + state.vel.x * seconds;
      out.y = state.pos.y + state.vel.y * seconds;
      out.z = state.pos.z + state.vel.z * seconds;
      return out;
    },
    reset(): void {
      accumulatorMs = 0;
      state.vel.x = 0;
      state.vel.y = 0;
      state.vel.z = 0;
    },
  };
}

File: client/Assets/Scripts/Sim/Predictor.cs
namespace Ac.Sim
{
    // C05 §5.4：预测骨架。50ms 定长子步进 + 渲染外推；和解重放留给后续计划。
    public sealed class Predictor
    {
        public const int SubstepMs = (int)LocalStep.StepDtMs;   // 与服务端 tick 同源，不允许出现第二个 50
        public const int MaxSubstepsPerFrame = 5;
        public const int MaxAccumulatorMs = 50;

        private MoveConfig _config;
        private int _accumulatorMs;

        public Predictor()
        {
            _config = MoveConfig.Default();
            State = new MoveState();
        }

        public MoveState State { get; private set; }
        public int AccumulatorMs { get { return _accumulatorMs; } }
        public int Steps { get; private set; }
        public int DroppedSubsteps { get; private set; }

        public void SetConfig(in MoveConfig config) { _config = config; }

        // §9：单命令步进（后续计划的和解重放共用这一条路径）。
        public bool StepWith(in StepCommand command)
        {
            var state = State;
            if (!LocalStep.StepLocalPlayer(ref state, command, _config, SubstepMs)) return false;
            State = state;
            Steps += 1;
            return true;
        }

        // §5.4：累加 dt，按 50ms 出步；一帧最多 5 步，超出丢弃余量并计 droppedSubsteps；
        // 余量上限 50ms（只保留一个子步），避免追帧雪崩。
        public int Advance(int dtMs, in StepCommand command)
        {
            if (dtMs <= 0) return 0;
            _accumulatorMs += dtMs;
            var stepped = 0;
            while (_accumulatorMs >= SubstepMs && stepped < MaxSubstepsPerFrame)
            {
                _accumulatorMs -= SubstepMs;
                StepWith(command);
                stepped += 1;
            }
            if (_accumulatorMs >= SubstepMs)
            {
                var dropped = _accumulatorMs / SubstepMs;
                DroppedSubsteps += dropped;
                _accumulatorMs = MaxAccumulatorMs;
            }
            return stepped;
        }

        // §5.4：renderPos = state.pos + state.vel * (accumulatorMs / 1000)
        public void RenderPosition(out double x, out double y, out double z)
        {
            var t = _accumulatorMs / 1000.0;
            x = State.X + State.Vx * t;
            y = State.Y + State.Vy * t;
            z = State.Z + State.Vz * t;
        }

        // 权威姿态写入（和解重放留给后续计划，这里只提供基线）。
        public void SetAuthoritative(double x, double y, double z, ushort yawUnits, ushort pitchUnits)
        {
            var state = State;
            state.X = x;
            state.Y = y;
            state.Z = z;
            state.YawRad = Quantize.DequantizeAngle(yawUnits);
            state.PitchRad = Quantize.DequantizeAngle(pitchUnits);
            State = state;
        }
    }
}
